### DISCUSSION TITLE
What is your preffered way of implementing permissions with Prisma?

### DISCUSSION THREAD

**Discussion by nenadfilipovic** [prisma/prisma#8524] [Q&A] [answered]
I currently have permissions implemented inside prisma query so I can allow update/delete/read operations based on some context.

Example:

```ts
where: {
AND: [ { some auth query like id: id from context },
{ business logic like name: "mike" } ]
```

Imagine this is query that is looking for user that have name mike and is owner of that account. I am composing queries this way and it is possible for every operation except creation because there is no "where" field when creating resource. Other problem is that with update and delete operations this is only possible with updateMany and deleteMany because those accept where field. Problem with this is that response will be BatchPayload and that will make problem with updating cache for graphlq client.

How do you guys handle authorization with prisma, I ideally look for a way where I can authorize and perform business logic in 1 request.

**Comment by ryands17** [comment, marked as answer]
@nenadfilipovic 👋 

I can't think of any way apart from using `updateMany`/`deleteMany` to keep it to a single query.

I personally use two queries, one for permissions and one for  the actual query. Take for e.g. this [snippet](https://github.com/ryands17/nexus-auth/blob/master/src/utils/rules.ts#L36-L50). What this does is create a authentication/authorisation handler for each Query/Mutation and that would execute before the resolver.

This will be equivalent to using middlewares in Express if you're familiar with those.

**Reply by nenadfilipovic** [reply]
Yea I am familiar with graphql-shield but I like to put my authentication and authorization low near models. I don't want to protect resolvers, instead when I protect resource I can use it anywhere and there is no need to rewrite it for every resolver.

Here is example for `update`:

```ts
export const updateDevice = async ({
  data,
  where,
  viewer,
}: UpdateDeviceArguments): Promise<NexusGenObjects['BatchPayload']> => {
  const ability = permissions(viewer);
  const composeQuery: Prisma.DeviceWhereInput[] = [];

  assert(data, deviceUpdateSchema);

  composeQuery.push(accessibleBy(ability, 'update').Device);
  composeQuery.push(where);

  return await prisma.device.updateMany({
    where: {
      AND: composeQuery,
    },
    data,
  });
};
```

I havent really checked but maybe adding unique fields can help, If I add `@unique[id, userId]` then I can use `update()` right because it is using `WhereDeviceUniqueInput`, two queries is always option.

**Reply by ryands17** [reply]
This would indeed work for updates and deletes, but for creates you would need two queries as you mentioned above.

**Reply by nenadfilipovic** [reply]
Yea Prisma does not have something like mongoose have to create resource but execute it when you want.

```ts
var new_user = new User({
    name: 'Manish',
    age:34
})
  
new_user.save()
```

If I had this in Prisma I could create resource, do a permission check and save it if it passes.

**Reply by ryands17** [reply]
Yeah Prisma isn't a class based ORM but we have a feature request for the same, so it would be great if you could add a 👍 to this and follow for more updates.
https://github.com/prisma/prisma/issues/5273

**Reply by nenadfilipovic** [reply]
Thanks for link, I added but it does not seem it will be available any time soon.

**Reply by ryands17** [reply]
Yes as per our [roadmap](https://www.notion.so/prismaio/Prisma-public-roadmap-50766227b779464ab98899accb98295f) not anytime soon.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
